evm_arithmetization/src/cpu/kernel/asm/memory/transient_storage.asm

```diff
@@ -0,0 +1,153 @@
+// Transient data storage
+
+
+/// The transient storage is stored in an array. The length of the array is stored in the global metadata.
+/// For storage keys, the address and key are stored as two consecutive elements.
+/// The array is stored in the SEGMENT_TRANSIENT_STORAGE segment in the kernel memory (context=0).
+/// Searching and inserting is done by doing a linear search through the array.
+/// If the key isn't found in the array, it is inserted at the end.
+/// TODO: Look into using a more efficient data structure.
+
+%macro search_transient_storage
+    %stack (addr, key) -> (addr, key, %%after)
+    %jump(search_transient_storage)
+%%after:
+    // stack:    (is_present, pos, addr, key, val)
+%endmacro
+
+/// Looks for an address, key pair into the transient storage. Returns 1 and the position in @SEGMENT_TRANSIENT_STORAGE
+/// if present or 0 and @GLOBAL_METADATA_TRANSIENT_STORAGE_LEN if not.
+global search_transient_storage:
+    // stack: addr, key, retdest
+    %mload_global_metadata(@GLOBAL_METADATA_TRANSIENT_STORAGE_LEN)
+    // stack: len, addr, key, retdest
+    PUSH @SEGMENT_TRANSIENT_STORAGE ADD
+    PUSH @SEGMENT_TRANSIENT_STORAGE
+search_transient_storage_loop:
+    // `i` and `len` are both scaled by SEGMENT_TRANSIENT_STORAGE
+    %stack (i, len, addr, key, retdest) -> (i, len, i, len, addr, key, retdest)
+    EQ %jumpi(search_transient_storage_not_found)
+    // stack: i, len, addr, key, retdest
+    DUP1
+    MLOAD_GENERAL
+    // stack: loaded_addr, i, len, addr, key, retdest
+    DUP4
+    // stack: addr, loaded_addr, i, len, addr, retdest
+    EQ
+    // stack: addr == loaded_addr, i, len, addr, key, retdest
+    DUP2
+    %increment
+    MLOAD_GENERAL
+    DUP6
+    EQ
+    MUL // AND
+    %jumpi(search_transient_storage_found)
+    // stack: i, len, addr, key, retdest
```

Comment:
I wonder if we couldn't do the loop in two stages to speed things up a bit. i.e., if addresses don't match, we start back from the beginning of the loop, and only load + check keys if they match.

That would be something like 

```asm
    // stack: i, len, addr, key, retdest
    DUP1
    MLOAD_GENERAL
    // stack: loaded_addr, i, len, addr, key, retdest
    DUP4
    // stack: addr, loaded_addr, i, len, addr, key, retdest
    SUB // functions as NEQ
    // stack: addr != loaded_addr, i, len, addr, key, retdest
    %jump(increment_and_recurse)

    // Addresses match, but we need to check for keys as well
    DUP2
    %increment
    MLOAD_GENERAL
    // stack: loaded_key, i, len, addr, key, retdest
    DUP5
    // stack: key, loaded_key, i, len, addr, key, retdest
    EQ
    %jumpi(search_transient_storage_found)
increment_and_recurse:
    // stack: i, len, addr, key, retdest
    %increment
    %jump(search_transient_storage_loop)
```

Reply:
thanks!